classes/user/Collector.inc.php

```diff
@@ -0,0 +1,589 @@
+        return $this;
+    }
+
+    /**
+     * Exclude results with the specified user IDs.
+     */
+    public function excludeUserIds(?array $excludeUserIds): self
+    {
+        $this->excludeUserIds = $excludeUserIds;
+        return $this;
+    }
+
+    /**
+     * Limit results to users enrolled in these roles
+     */
+    public function filterByRoleIds(?array $roleIds): self
+    {
+        $this->roleIds = $roleIds;
+        return $this;
+    }
+
+    /**
+     * Limit results to users enrolled in these roles
+     */
+    public function filterByWorkflowStageIds(?array $workflowStageIds): self
+    {
+        $this->workflowStageIds = $workflowStageIds;
+        return $this;
+    }
+
+
+    /**
+     * Limit results to users with user groups in these context IDs
+     */
+    public function filterByContextIds(?array $contextIds): self
+    {
+        $this->contextIds = $contextIds;
+        return $this;
+    }
+
+    /**
+     * Limit results to users registered before a specified date.
+     */
+    public function filterRegisteredBefore(?string $registeredBefore): self
+    {
+        $this->registeredBefore = $registeredBefore;
+        return $this;
+    }
+
+    /**
+     * Limit results to users registered after a specified date.
+     */
+    public function filterRegisteredAfter(?string $registeredAfter): self
+    {
+        $this->registeredAfter = $registeredAfter;
+        return $this;
+    }
+
+    public function includeReviewerData(bool $includeReviewerData = true): self
+    {
+        $this->includeReviewerData = $includeReviewerData;
+        return $this;
+    }
+
+    /**
+     * Retrieve a set of users not assigned to a given submission stage as a user group.
+     * (Replaces UserStageAssignmentDAO::getUsersNotAssignedToStageInUserGroup)
+     */
+    public function filterExcludeSubmissionStage(int $submissionId, int $stageId, int $userGroupId): self
+    {
+        $this->excludeSubmissionStage = [
+            'submission_id' => $submissionId,
+            'stage_id' => $stageId,
+            'user_group_id' => $userGroupId,
+        ];
+        return $this;
+    }
+
+    /**
+     * Retrieve StageAssignments by submission and stage IDs.
+     * (Replaces UserStageAssignmentDAO::getUsersBySubmissionAndStageId)
+     */
+    public function filterSubmissionAssignment(?int $submissionId, ?int $stageId = null, ?int $userGroupId = null): self
+    {
+        if ($submissionId === null) {
+            // Clear the condition.
+            $this->submissionAssignment = null;
+            if ($stageId !== null || $userGroupId !== null) {
+                throw new \InvalidArgumentException('If a stage or user group ID is specified, a submission ID must be specified as well.');
+            }
+        } else {
+            $this->submissionAssignment = [
+                'submission_id' => $submissionId,
+                'stage_id' => $stageId,
+                'user_group_id' => $userGroupId,
+            ];
```

Comment:
I kept these three together because they are intertwined and refer to a submission assignment. We already have a `filterByUserGroupId` (and `userGroupIds` property) separate from the one here, used to check user group enrollment (as opposed to assignment to a submission with a particular user group). Likewise, we have another join on `stage_assignments` for excluded submission stages. I suppose I could use `assignedSubmissionId`, `assigedSubmissionStageId`, and `assignedSubmssionUserGroupId`, but I'm not sure that's clearer.